rfc/rfc017-distributed-network-grpc-v2.md

```diff
@@ -45,4 +54,270 @@ As specified by RFC015, the node MUST authenticate the peer's node DID as follow
+Although the set reconciliation protocol can also synchronize any difference in transactions sets between nodes, it's not very efficient for small differences.
+
+### 5.1 Data requirements
+
+#### 5.1.1 XOR
+
+The exclusive-or (XOR) value is calculated over all transaction references within a range.
+The size of the XOR value is therefore the same as the transaction reference size (SHA256: 32 bytes).
+The XOR value is used to quickly determine if two nodes have processed the exact same set of transactions.
+
+### 5.2 Operation
+
+All messages and values mentioned in this chapter are scoped to a single connection between two nodes.
+*Alice* and *Bob* are used to represent two nodes connected to each other.
+
+The protocol generally operates as follows:
+
+1. Alice sends at a set interval (§5.2.1):
+   * XOR of all known transaction references.
+   * The highest LC value.
+   * A list of transaction references since the previous message. If no previous message has been send, the list is empty.
+
+2. When receiving Alice's message, Bob compares the XOR value from Alice with its own (§5.2.2):
+   * When the XOR is the same: no action required.
+   * When different:
+      * Remove all known transaction references from the list.
+      * Request the remaining transactions.
+      * If the remaining list is empty and the LC value is equal or higher than Bob's highest LC value, Bob sends a `State` message as defined in Chapter 6. 
+
+3. Alice responds with the requested transactions (§5.2.3).
+
+4. After adding Alice's transactions to the DAG (making sure its cryptographic signature is valid), Bob SHOULD query the payload if it's missing.
+
+A node MUST make sure to only add transactions of which all previous transactions are present. If previous transactions are missing, Bob will send a `State` message.
+
+#### 5.2.1 Broadcasting new transactions
+
+A node's new transaction references MUST be broadcast at an interval using the `Gossip` message, by default every 2 seconds. The interval MAY be adjusted by the node operator but MUST conform to the limits (min/max interval) defined by the network. It is advised to keep it relatively short because it directly influences the speed by which new transactions are retrieved.
+
+The `Gossip` message MUST contain an `XOR` value, an `LC` value and a list of transaction references (`transactions`).
+The list MUST contain all new transaction references since the last `Gossip` message.
+The list of transaction references MUST be tracked per connection.
+If no new transactions have been added/received or for the first message for a connection, an empty list is sent.
+The list MUST not contain more than 100 transactions.
+This could create a backlog of messages at the sending node's side. 
+A node SHOULD take precautions on keeping this backlog below maximum, eg it SHOULD reduce the inflow of new transactions.
```

Comment:
That will have different consequences in different situations. You can't prefer your own TXs since they might depend on others previous transactions. It's also a SHOULD, not a MUST so it's up to the implementing party.

A restriction on inflow could be to pause requesting TXs from other nodes to give way to own TXs. If that is too much as well, a restriction on the API should occur. Also a buffer could be added in between allowing for some bursts of TXs. Then later when things cool down, messaging can catch up.

Reply:
Restricting a node's intake of new Txs could cause the node to fall behind and potentially never catch up with the rest of the network. This could introduce a delay in authorization of access to patient data. I would prefer to reduce the interval for gossip messages in this situation.

The current requirements also forces all nodes, including 'spectators', to be an active participant in distribution of information. Generally speaking this is a good thing, but it might not be feasible for all to do so.

The exact gossip message probably requires some more thought. I prefer the message to have a minimal set of requirements so the exact form can be adapted when we learn more. I think the gossip messages should be more of the broadcast and forget type. It doesn't feel right that this has any impact on how/if the node accepts/creates new Txs.

Reply:
```suggestion
The list MUST not contain more than 100 transactions.
This could create a backlog of messages at the sending node's side. 
A node SHOULD take precautions to keep this backlog to a minimum.
```